File: visual-python/src/saucelabs_visual/regions.py
from typing import Union, List


class Region:
    name: str
    aliases: List[str] = []
    graphql_endpoint: str
    job_url_template: str
    visual_url_template: str

    def __init__(
            self,
            name: str,
            aliases: List[str],
            graphql_endpoint: str,
            job_url_template: str,
            visual_url_template: str,
    ):
        self.name = name
        self.aliases = aliases
        self.graphql_endpoint = graphql_endpoint
        self.job_url_template = job_url_template
        self.visual_url_template = visual_url_template

    def job_url(self, job_id: str) -> str:
        return self.job_url_template.format(JOB_ID=job_id)

    def build_url(self, build_id: str) -> str:
        return self.visual_url_template.format(BUILD_ID=build_id)

    @staticmethod
    def from_name(region_name: str) -> 'Region':
        region: Union[Region, None] = None

        for i, v in enumerate(regions):
            if v.name == region_name or region_name in v.aliases:
                region = v
                break

        if region is None:
            raise Exception(f'"{region_name}" is not recognized as a Sauce Labs region.')

        return region


regions = [
    Region(
        name="staging",
        aliases=["us-west-4-jeh6"],
        graphql_endpoint='https://api.staging.saucelabs.net/v1/visual/graphql',
        job_url_template='https://app.staging.saucelabs.net/tests/{JOB_ID}',
        visual_url_template='https://app.saucelabs.com/visual/builds/{BUILD_ID}',
    ),
    Region(
        name="us-east-4",
        aliases=["us-east-4-cm5i"],
        graphql_endpoint='https://api.us-east-4.saucelabs.com/v1/visual/graphql',
        job_url_template='https://app.us-east-4.saucelabs.com/tests/{JOB_ID}',
        visual_url_template='https://app.us-east-4.saucelabs.com/visual/builds/{BUILD_ID}',
    ),
    Region(
        name="eu-central-1",
        aliases=["eu", "eu-west-3-lnbf"],
        graphql_endpoint='https://api.eu-central-1.saucelabs.com/v1/visual/graphql',
        job_url_template='https://app.eu-central-1.saucelabs.com/tests/{JOB_ID}',
        visual_url_template='https://app.eu-central-1.saucelabs.com/visual/builds/{BUILD_ID}',
    ),
    Region(
        name="us-west-1",
        aliases=['us', 'us-west-4-i3er'],
        graphql_endpoint='https://api.us-west-1.saucelabs.com/v1/visual/graphql',
        job_url_template='https://app.saucelabs.com/tests/{JOB_ID}',
        visual_url_template='https://app.saucelabs.com/visual/builds/{BUILD_ID}',
    ),
]

File: visual-js/visual/src/regions.ts
import { SauceRegion } from './api';

export class VisualApiRegion {
  name: SauceRegion;
  aliases: string[];
  graphqlEndpoint: string;
  jobUrlTemplate: string;

  constructor(options: {
    name: SauceRegion;
    aliases: string[];
    graphqlEndpoint: string;
    jobUrlTemplate: string;
  }) {
    this.name = options.name;
    this.aliases = options.aliases;
    this.graphqlEndpoint = options.graphqlEndpoint;
    this.jobUrlTemplate = options.jobUrlTemplate;
  }

  jobUrl(jobId: string): string {
    return this.jobUrlTemplate.replace('JOB_ID', jobId);
  }

  static fromName(name: string) {
    const r = regions.find((r) => r.name === name || r.aliases.includes(name));
    if (!r) {
      throw new Error(`${name} is not recognised as a Sauce Labs region.`);
    }
    return r;
  }
}

export const regions: VisualApiRegion[] = [
  new VisualApiRegion({
    name: 'staging',
    aliases: ['us-west-4-jeh6'],
    graphqlEndpoint: 'https://api.staging.saucelabs.net/v1/visual/graphql',
    jobUrlTemplate: 'https://app.staging.saucelabs.net/tests/JOB_ID',
  }),
  new VisualApiRegion({
    name: 'us-east-4',
    aliases: ['us-east-4-cm5i'],
    graphqlEndpoint: 'https://api.us-east-4.saucelabs.com/v1/visual/graphql',
    jobUrlTemplate: 'https://app.us-east-4.saucelabs.com/tests/JOB_ID',
  }),
  new VisualApiRegion({
    name: 'eu-central-1',
    aliases: ['eu', 'eu-west-3-lnbf'],
    graphqlEndpoint: 'https://api.eu-central-1.saucelabs.com/v1/visual/graphql',
    jobUrlTemplate: 'https://app.eu-central-1.saucelabs.com/tests/JOB_ID',
  }),
  new VisualApiRegion({
    name: 'us-west-1',
    aliases: ['us', 'us-west-4-i3er'],
    graphqlEndpoint: 'https://api.us-west-1.saucelabs.com/v1/visual/graphql',
    jobUrlTemplate: 'https://app.saucelabs.com/tests/JOB_ID',
  }),
];

export const resolveRegionFromOndemandHostname = (
  hostname: string,
): string | undefined => {
  const match = hostname.match(
    /^ondemand\.(?:([a-z0-9-]+)\.)?saucelabs\.(?:com|net)$/,
  );
  if (!match) {
    return undefined;
  }
  if (!match[1]) {
    return 'us-west-1';
  }
  return match[1];
};
